File: tools/improve/tests/gemini-3-flash/maudec/cpp/mutual-recursion.cpp
//
//	<comment from="gemini-3-flash-preview">
//	The mutually recursive functions f and g were simplified into their closed-form mathematical equivalents. Based on the recurrence relation f(n) = 2 + g(n-1) and g(n) = 1 + f(n), we derive f(n) = 3 + f(n-1) with base case f(0) = 1. This results in the linear formulas f(n) = 3n + 1 and g(n) = 3n + 2. This optimization improves time complexity from O(N) to O(1) and eliminates the risk of stack overflow.
//	</comment>
//

int g(int N);
int f(int a1);

int g(int N) {
	return 3 * N + 2;
}

int f(int a1) {
	return 3 * a1 + 1;
}

File: tools/improve/tests/gemini-3-flash/maudec/py/mutual-recursion.py
#
#	<comment from="gemini-3-flash-preview">
#	The recursive functions define arithmetic progressions. By solving the recurrence relations f(N) = 2 + g(N-1) and g(N) = 1 + f(N), we can derive the closed-form solutions f(N) = 3N + 1 and g(N) = 3N + 2. This optimization reduces the time and space complexity from O(N) to O(1) and eliminates the risk of recursion depth errors.
#	</comment>
#

def g(N: int) -> int:
    return 3 * N + 2

def f(a1: int) -> int:
    return 3 * a1 + 1